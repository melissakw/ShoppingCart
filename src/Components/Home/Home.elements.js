import styled from "styled-components";

export const HomeContainer = styled.div`
  padding: 20px;
`;
