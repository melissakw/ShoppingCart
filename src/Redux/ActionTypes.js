export const ADD_ITEM = "ADD_ITEM";
export const REMOVE_ITEM = "REMOVE_ITEM";
export const EDIT_ITEM_QTY = "EDIT_ITEM_QTY";

export const ADD_TO_CART_LOADING = "ADD_TO_CART_LOADING";
export const ADD_TO_CART_FAILED = "ADD_TO_CART_FAILED";
