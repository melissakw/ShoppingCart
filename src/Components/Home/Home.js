import React, { useEffect, useState } from "react";
import ProductsPage from "../ProductsPage/ProductsPage";
import HeaderBar from "../HeaderBar/HeaderBar";
import axios from "axios";
import CircularProgress from "@material-ui/core/CircularProgress";
import { HomeContainer } from "./Home.elements";
import { AppContainer, LoadingContainer } from "../App.elements";

const Home = () => {
  const [products, setProducts] = useState([]);
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState(null);
  const endpoint = "http://localhost:8000/products";

  const fetchProducts = async () => {
    setIsLoading(true);
    setError(null);

    try {
      const response = await axios.get(endpoint);
      let updatedResponse = response.data.map((obj) => {
        delete obj["description"];
        delete obj["category"];
        return obj;
      });
      setProducts(updatedResponse);
    } catch (e) {
      setError(e);
    }
    setIsLoading(false);
  };

  useEffect(() => {
    fetchProducts();
  }, []);

  return (
    <AppContainer>
      <HeaderBar />

      <HomeContainer>
        <h1>Products</h1>
        {isLoading && (
          <LoadingContainer>
            <CircularProgress />
          </LoadingContainer>
        )}
        <div>{error && <p>Oops... please try again!</p>}</div>
        <ProductsPage products={products} />
      </HomeContainer>
    </AppContainer>
  );
};

export default Home;
