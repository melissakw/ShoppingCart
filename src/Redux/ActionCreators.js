import * as ActionTypes from "./ActionTypes";

export const postToCart = (product) => (dispatch) => {
  const productWithQty = {
    id: product.id,
    title: product.title,
    price: product.price,
    image: product.image,
    quantity: 1, //defualt quantity to 1 when adding item to cart
    stock: 10
  };
  dispatch(addToCart(productWithQty));
};

export const addToCart = (product) => ({
  type: ActionTypes.ADD_ITEM,
  payload: product
});

export const removeItem = (id) => ({
  type: ActionTypes.REMOVE_ITEM,
  payload: id
});

export const editItem = (product, editedQuantity) => ({
  type: ActionTypes.EDIT_ITEM_QTY,
  product: product,
  quantity: editedQuantity
});

export const addToCartLoading = () => ({
  type: ActionTypes.ADD_TO_CART_LOADING
});

export const addToCartFailed = (errmess) => ({
  type: ActionTypes.ADD_TO_CART_FAILED,
  payload: errmess
});
