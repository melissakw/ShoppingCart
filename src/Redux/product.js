import * as ActionTypes from "./ActionTypes";

export const product = (
  state = {
    isLoading: true,
    errMess: null,
    cartItems: []
  },
  action
) => {
  switch (action.type) {
    case ActionTypes.ADD_ITEM:
      let duplicateItem = false;
      let copyItems = state.cartItems.map((item) => {
        if (item.id === action.payload.id) {
          item.quantity = Math.min(item.quantity + 1, item.stock);
          duplicateItem = true;
        }
        return item;
      });

      if (duplicateItem) {
        return { ...state, cartItems: copyItems };
      } else {
        return {
          ...state,
          cartItems: [action.payload].concat(state.cartItems)
        };
      }

    case ActionTypes.REMOVE_ITEM:
      return {
        ...state,
        cartItems: state.cartItems.filter((item) => item.id !== action.payload)
      };
    case ActionTypes.EDIT_ITEM_QTY:
      let updatedProductWithNewQty = state.cartItems.map((item) => {
        if (item.id === action.product.id) {
          item.quantity = action.quantity;
        }
        return item;
      });
      return { ...state, cartItems: updatedProductWithNewQty };

    default:
      return state;
  }
};
