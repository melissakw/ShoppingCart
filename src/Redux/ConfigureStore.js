import { createStore, applyMiddleware } from "redux";
import { persistStore, persistCombineReducers } from "redux-persist";
import storage from "redux-persist/lib/storage";
import thunk from "redux-thunk";
import logger from "redux-logger";

import { product } from "./product";

export const ConfigureStore = () => {
  const config = {
    key: "root",
    storage: storage,
    debug: true
  };

  const store = createStore(
    persistCombineReducers(config, {
      product
    }),
    applyMiddleware(thunk, logger)
  );

  const persistor = persistStore(store);
  return { persistor, store };
};
