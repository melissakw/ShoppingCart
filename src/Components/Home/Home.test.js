import React from 'react';
import Home from './Home'
import '../../test.json';
import axiosMock from 'axios';
import renderer from 'react-test-renderer';
import { render,cleanup, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { Provider } from "react-redux";
import { ConfigureStore } from "../../Redux/ConfigureStore";
import { BrowserRouter } from 'react-router-dom';

beforeEach(() => axiosMock.get.mockResolvedValueOnce({data: require('../../test.json')})
);
afterEach(cleanup);

test("renders Home correctly", () => {

	const { store } = ConfigureStore();
	const tree = renderer.create(
		<Provider store={store}>
			<BrowserRouter>
				<Home />
			</BrowserRouter>
		</Provider>
	).toJSON();

	expect(tree).toMatchSnapshot();
});

test("sucessfully fetches and renders data", async() => {
	
	const { store } = ConfigureStore();
	const {getByText} = render(
		<Provider store={store}>
			<BrowserRouter>
				<Home />
			</BrowserRouter>
		</Provider>
	);
	
	expect(await screen.findByText(/SanDisk SS/i)).toBeInTheDocument();

});
