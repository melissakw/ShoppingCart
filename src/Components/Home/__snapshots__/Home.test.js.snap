// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Home correctly 1`] = `
<div
  className="sc-gKsewC bBvTIN"
>
  <div
    className="sc-iBPRYJ jYpaRc"
  >
    <a
      className="sc-fubCfw hIQLhz"
      href="/"
      onClick={[Function]}
    >
      <h3>
        THE SHOP
      </h3>
    </a>
    <div
      className="sc-pFZIQ JNXgK"
    >
      <a
        href="/cart"
        onClick={[Function]}
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          data-testid="shopping-icon"
          focusable="false"
          style={
            Object {
              "color": "black",
            }
          }
          viewBox="0 0 24 24"
        >
          <path
            d="M7 18c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zM1 2v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.14 0-.25-.11-.25-.25l.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.58-6.49c.08-.14.12-.31.12-.48 0-.55-.45-1-1-1H5.21l-.94-2H1zm16 16c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2z"
          />
        </svg>
      </a>
      <div
        className="sc-jrAGrp gBlpaS badge badge-dark"
      >
        0
      </div>
    </div>
  </div>
  <div
    className="sc-iqHYGH kUloaI"
  >
    <h1>
      Products
    </h1>
    <div />
    <div>
      <div
        className="sc-bdfBwQ pdzVR"
      />
    </div>
  </div>
</div>
`;
